coalib/core/Bear.py

```diff
@@ -0,0 +1,458 @@
+        # exist also as fields.
+        excluded_members = {'can_detect', 'maintainers', 'maintainers_emails'}
+
+        # json cannot serialize properties, so drop them
+        data = {
+            key: value
+            for key, value in get_public_members(cls).items()
+            if not isinstance(value, property) and key not in excluded_members}
+
+        metadata = cls.get_metadata()
+        non_optional_params = metadata.non_optional_params
+        optional_params = metadata.optional_params
+        data['metadata'] = {
+            'desc': metadata.desc,
+            'non_optional_params': {param: non_optional_params[param][0]
+                                    for param in non_optional_params},
+            'optional_params': {param: optional_params[param][0]
+                                for param in optional_params}}
+
+        return data
+
+    @staticmethod
+    def setup_dependencies():
+        """
+        This is a user defined function that can download and set up
+        dependencies (via download_cached_file or arbitrary other means) in an
+        OS independent way.
+        """
+
+    @classmethod
+    def check_prerequisites(cls):
+        """
+        Checks whether needed runtime prerequisites of the bear are satisfied.
+
+        This function gets executed at construction.
+
+        Section value requirements shall be checked inside the ``run`` method.
+
+        >>> from dependency_management.requirements.PipRequirement import (
+        ...     PipRequirement)
+        >>> class SomeBear(Bear):
+        ...     REQUIREMENTS = {PipRequirement('pip')}
+
+        >>> SomeBear.check_prerequisites()
+        True
+
+        >>> class SomeOtherBear(Bear):
+        ...     REQUIREMENTS = {PipRequirement('really_bad_package')}
+
+        >>> SomeOtherBear.check_prerequisites()
+        'Following requirements are not installed: really_bad_package (...)'
+
+        :return: True if prerequisites are satisfied, else False or a string
+                 that serves a more detailed description of what's missing.
+        """
+        not_installed_requirements = [requirement
+                                      for requirement in cls.REQUIREMENTS
+                                      if not requirement.is_installed()]
+
+        if not_installed_requirements:
+            return 'Following requirements are not installed: ' + ', '.join(
+                '{} (installable via `{}`)'.format(
+                    requirement.package,
+                    ' '.join(requirement.install_command()))
+                for requirement in not_installed_requirements)
+        else:
+            return True
+
+    def get_config_dir(self):
+        """
+        Gives the directory where the configuration file resides.
+
+        :return:
+            Directory of the config file.
+        """
+        return get_config_directory(self.section)
+
+    @classmethod
+    def download_cached_file(cls, url, filename):
```

Comment:
@jayvdb just added some changes to this method in another PR on the current Bear.py you might want to incorporate them. They do address some bugs.